gax/src/main/java/com/google/api/gax/rpc/mtls/MtlsUtils.java

```diff
@@ -0,0 +1,178 @@
+      @Override
+      public String getenv(String name) {
+        return System.getenv(name);
+      }
+    }
+
+    DefaultMtlsProvider() {
+      this(new SystemEnvironmentProvider(), DEFAULT_CONTEXT_AWARE_METADATA_PATH);
+    }
+
+    private EnvironmentProvider envProvider;
+    private String metadataPath;
+
+    /** Cache the key store so we don't create a new one every time getKeyStore is called. */
+    private KeyStore keyStore = null;
+
+    @VisibleForTesting
+    DefaultMtlsProvider(EnvironmentProvider envProvider, String metadataPath) {
+      this.envProvider = envProvider;
+      this.metadataPath = metadataPath;
+    }
+
+    @Override
+    public boolean useMtlsClientCertificate() {
+      String useClientCertificate = envProvider.getenv(GOOGLE_API_USE_CLIENT_CERTIFICATE);
+      return "true".equals(useClientCertificate);
+    }
+
+    @Override
+    public String useMtlsEndpoint() {
+      return envProvider.getenv(GOOGLE_API_USE_MTLS_ENDPOINT);
+    }
+
+    @Override
+    public String getKeyStorePassword() {
+      return "";
+    }
+
+    @Override
+    public KeyStore getKeyStore() throws IOException, GeneralSecurityException {
+      if (keyStore != null) {
+        // return the cached key store
+        return keyStore;
+      }
+      try {
+        // Load the cert provider command from the json file.
+        InputStream stream = new FileInputStream(metadataPath);
+        List<String> command = extractCertificateProviderCommand(stream);
+
+        // Run the command and timeout after 1000 milliseconds.
+        Process process = new ProcessBuilder(command).start();
+        int exitCode = runCertificateProviderCommand(process, 1000);
+        if (exitCode != 0) {
+          throw new IOException("Cert provider command failed with exit code: " + exitCode);
+        }
+
+        // Create mTLS key store with the input certificates from shell command.
+        keyStore = SecurityUtils.createMtlsKeyStore(process.getInputStream());
+        return keyStore;
+      } catch (FileNotFoundException ignored) {
+        // file doesn't exist
+        return null;
+      } catch (InterruptedException e) {
+        throw new IOException("Interrupted executing certificate provider command", e);
+      }
+    }
+
+    @VisibleForTesting
+    static List<String> extractCertificateProviderCommand(InputStream contextAwareMetadata)
+        throws IOException {
+      JsonParser parser = new JacksonFactory().createJsonParser(contextAwareMetadata);
+      ContextAwareMetadataJson json = parser.parse(ContextAwareMetadataJson.class);
+      return json.getCommands();
+    }
+
+    @VisibleForTesting
+    static int runCertificateProviderCommand(Process commandProcess, long timeoutMilliseconds)
+        throws IOException, InterruptedException {
+      long startTime = System.currentTimeMillis();
+      long remainTime = timeoutMilliseconds;
+      boolean terminated = false;
+
+      do {
```

Comment:
It seems like it makes sence to wait a little bit before doing the first attempt, because most likely the first attempt will always throw `IllegalThreadStateExcpeiton`. Would it make sense to add some waiting before the first check, and/or, as a result, convert this do-while loop into while?

Reply:
Now converted to a while loop.